gcc 11 warning: null destination pointer

**Describe the bug**
When compiling with ~~clang 18~~ GCC 11 and optimisations turned on:
```
In function ‘int sprintf(char*, const char*, ...)’,
    inlined from ‘char* quote_block(const void*, int)’ at /home/louis/Personal/freeciv21/server/savegame/savegame3.cpp:837:10,
    inlined from ‘void sg_save_player_attributes(savedata*, player*)’ at /home/louis/Personal/freeciv21/server/savegame/savegame3.cpp:6502:31,
    inlined from ‘void sg_save_players(savedata*)’ at /home/louis/Personal/freeciv21/server/savegame/savegame3.cpp:3823:30:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:30:34: warning: null destination pointer [-Wformat-overflow=]
   30 |   return __builtin___sprintf_chk (__s, __USE_FORTIFY_LEVEL - 1,
      |          ~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   31 |                                   __glibc_objsize (__s), __fmt,
      |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   32 |                                   __va_arg_pack ());
      |                                   ~~~~~~~~~~~~~~~~~
```

**To Reproduce**
Steps to reproduce the behavior:
1. Use Ubuntu 24.04 and clang
3. Turn on optimizations
4. Build
5. See error

**Expected behavior**
No warning

**Platform and version (please complete the following information):**
 - OS: Ubuntu 24.04
 - Freeciv21 version: master

**Additional context**
Add any other context about the problem here.

--- a/server/savegame/savegame3.cpp
+++ b/server/savegame/savegame3.cpp
@@ -825,20 +825,20 @@ static enum unit_activity char2activity(char activity)
 
 /**
    Quote the memory block denoted by data and length so it consists only of
-   " a-f0-9:". The returned string has to be freed by the caller using
- free().
+   " a-f0-9:". The returned string has to be deleted by the caller.
  */
 static char *quote_block(const void *const data, int length)
 {
-  char *buffer = new char[length * 3 + 10];
+  auto bufsize = length * 3 + 10;
+  char *buffer = new char[bufsize];
   size_t offset;
   int i;
 
-  sprintf(buffer, "%d:", length);
+  snprintf(buffer, bufsize, "%d:", length);
   offset = qstrlen(buffer);
 
   for (i = 0; i < length; i++) {
-    sprintf(buffer + offset, "%02x ", ((unsigned char *) data)[i]);
+    snprintf(buffer + offset, bufsize, "%02x ", ((unsigned char *) data)[i]);
     offset += 3;
   }
   return buffer;

